### ISSUE TITLE
meta-properties and multi-properties for security and history 

### ISSUE THREAD

**Issue by BrynCooke** [tinkerpop/tinkerpop3#489]
Recently I have been looking at history and security GraphStrategies with @rjbriody and it seems like meta-properties and multi-properties may not fulfil their potential in their current form. In particular due to the lack of support on edges.

Meta and multi-property support is at the cost of a more complex element api than TP2 get/setProperty, so it's worth making sure that the extra effort is worth it for end users. (https://github.com/tinkerpop/tinkerpop3/issues/389)

Here are the use cases and the difficulties faced:

### SecurityGraphStrategy
* User wants to restrict access to reading/writing a property on an element on the graph. This applies equally edges as well as vertices.

#### Issue
* ACL information could be stored in meta-properties. However as there is no support on edges I end up using prefixed properties instead.
* Lists of users and groups could be store in multi-properties, but again the lack of edge support means that I end up concatenating the users and groups to a string which is then indexed via contains.

There seems little point in having a different strategy for edges and vertices, so currently I'm not using VertexProperty at all to implement security.

### HistoryGraphStrategy
* User wants to be able to view the state of an element at any point in history. This applies to edges as well as vertices.

#### Issue
* Historical property data could be stored in a list in a meta property. However, edges are not supported and meta properties can not also have meta properties, so if the original property had meta properties they could not be stored in a met-property.
* Historical property data could be stored in a list in a prefixed property. However, edges are not supported. I will probably end up creating a history vertex and storing the data there. Edges will have to be linked via UUID to the history vertex and for consistency vertices will probably do the same. This doesn't feel like a nice solution.

### Observations
* I'm having to fall back to prefixing way too often.
* The benefit of meta-properties does not seem to be clear cut over prefixing especially as they are not supported on edges, but also that meta-properties can also not have meta-properties.
* End users will not understand why there is a distinction between element types for properties, and if they have to support edges then they'll end up using prefixing and custom serialization anyway. No matter how many times the reasoning is explained people will continue to ask. Get ready to start pushing a boulder up a hill.

And lastly a question:
If via custom serialization e.g. JSON I can store structured data in a property and it will work on all types of element, then is a dedicated API only on vertices definitely worth the extra complexity? If I am really lucky then my Graph implementation may even support indexing JSON.

**Comment by BrynCooke** [comment]
I think the paradox is that multi and meta properties could make it easy to create history on tp2 graphs but cannot represent themselves in history.

**Comment by davidcrouch** [comment]
I'm doing something almost identical (AuthorizationStrategy, AuditStrategy ported from TP2 wrappers) and am facing the same issues because of no meta-properties and multi-properties on edges. The benefits of vertex properties are negated with the constraints of edge properties and like Bryn I'm being forced to develop to the lowest common denominator (which is almost back to TP2 patterns). I'm trying to develop EdgeMetaPropertyStrategy as a workaround (and struggling).

Given limitations of a vendor graph DB, creative graph strategy developers will develop workarounds to fill the gaps (in the form of combinations of vertices, edges and properties). I'm finding that the TP3 data model (with new vertex property concept) is deceiving because my workarounds now need to consider both the vertex property and property concepts (added complexity). 

I think some focus needs to be put into the extensibility of TP3 that encourage developers to create workarounds for graph DB limitations using TP3 vs creating workarounds for TP3. I'm happy to provide some cycles to help.

**Comment by rjbriody** [comment]
+1 

I have already seen this a few times around ACL/Permissions. Users expect to treat Edges and Vertices as symmetric first-class citizens. "forced to develop to the lowest common denominator" is an accurate way to put it.

I understand that there are vendor storage limitations that make this hard. I doubt we would have this design if it weren't for these limitations. I am +1'ing anyway in the hope that there may still be a creative solution out there.

**Comment by davidcrouch** [comment]
Bob,

I was approaching it a little differently. Rather than force requirements upon vendors, which they may not be able to easily implement, provide a base implementation as it is now but also provide the extensibility hooks and tools in TP3 for a developer to code custom extensions. For example, a developer could write a strategy that uses a vertex (which obviously supports meta-properties in the underlying database) to represent the edge concept. It could look like this:-

interface MyEdge extends Edge<VertexProperty>, Vertex, Vertex.<VertexProperty>Iterators

........  then be able to use MyEdge wherever an Edge can be used but with the extra features of a vertex (meta properties AND hyper-edges for free - g.V().outE().out())

Improved extensibility would encourage the creativity of developers and is not a burden on the vendors. It would be pure TP3 and hence continue the charter of promoting vendor neutrality.

**Comment by mbroecheler** [comment]
I think it is a mistake to treat vertices and edges as symmetric first class citizens and we need to be careful to get away from it. It's true that the original property graph model encouraged that point of view, but if you logically pursue that approach to graph data models it leads you astray.
For the sake of illustration, assume we consider vertices and edges symmetric first class citizen. Then, what's the difference between the two? Well, an edge connects vertices. In other words, an edge is a vertex which points to an in and out vertex. The question then is: why even make that distinction and why limit it to one in and out vertex? Wouldn't hyper-edges be preferable so you could model a relationship between a seller, a buyer, and a product? In fact, that is essentially Codd's argument against the network data model. He argued that the distinction was arbitrary and should be removed in favor of just one first class citizen: the relation. And a relation can have foreign key constraints to link it to other relations.

Does that mean the graph data model is superfluous and we will eventually go back to the relational model? I believe there is a strong argument to be made in favor of the graph data model if - and only if - we clearly differentiate between vertices and edges.
Vertices represent entities in a domain and edges represent a relationship between pairs of entities. So, first class: vertex. Second class: edge. Vertex properties are also second class citizens as they represent a relationship between a vertex and a value (again, a binary relationship). And there are third class citizens: key-value annotations on vertex-properties and edges.

As you can see, the graph data model has more going on than the relational data model which can be more confusing, but I think that's also a great strength of the graph data model because it forces the user to break down their domain into binary relationships which is close to our mental model and therefore makes data modeling easier and more productive.
However, it is important to stress the distinction between vertices and edges. Otherwise we end up undermining the graph data model because people will do things like model the above mentioned transaction as an edge between buyer and seller and then slap the product as a property on the edge. The challenge is to get ourselves away from the relational model. I believe a lot of folks coming to graphs think of vertices and edges as relations and that gets them into trouble in the medium to long term.

Your entities should be vertices, your binary relationship edges and vertex-properties. That's the starting point. If you are adding properties to edges at this high level stage of data modeling, they should become vertices. When adding properties to edges and vertex-properties, they should be qualifying not identifying. If you add an order-no to a transaction edge, most likely that edge should be a vertex.
Isn't the distinction between entities and relationships somewhat arbitrary? Somewhat. I think a fatal flaw of data modeling (due to historic constraints) is to think of data modeling as something static - you are trying to capture this entire world perfectly right. That is impossible. But at any one point in time, your understanding of the domain combined with the application requirements distill down to what entities and relationships are. Those are obviously changing as your application evolves and I think we will see a lot more tool support around that in graph databases - i.e. going from edges to vertices - and there will be the equivalent of view maintenance which are path contractions in graphs. That said, however, the first question of graph data modeling should always be: what are my entities and what are the binary relationships (to other entities and values) they are engaged in.

Now, let's break this back down to ACL/provenance/history graph wrappers. By the logic above, if you were to track changes to individual properties on an edge, that edge is treated as an entity and should be a vertex. If you are changing a property, you are changing the qualification of the edge, thereby effectively (from a logical data model point of view) creating a new edge. So, a common approach would be to soft-delete the old edge and add a new edge with updated provenance information on it.

**Comment by davidcrouch** [comment]
Hi Matthias,

Firstly, thank you for you detailed post. Secondly, I fundamentally agree with you. However, I want to point out two things, one from a philosophical point of view and the second from a practical point of view. 

Regarding your perspective on the separation of the graph elements (vertex, edge  etc.), I 100% agree with you in that its important to separate these concepts to preserve the integrity of the graph data model. However, that's what it is - a data model. As you move up to higher level abstractions the meaning of these concepts evolves. Now, I've been using Blueprints for a number of years now and did exactly what you pointed out when I implemented authorization, version control etc. Each property of a "vertex" had its own vertex and each value for a property had its own vertex. We built it all in TP2 with multiple wrapper layers. It works and its speed is acceptable. We also went a little RDF'y and represented edges as properties on a vertex. And with this perspective I totally understand the similarities of vertex properties and their properties and how they are similar to edges and their properties. 

But when you start to move to higher abstraction levels - more to object modeling, the relevance of ACLs etc. becomes a distraction and just noise. It's extremely hard to a debug a business level traversal when you're deciphering vertices used for permissions. Given this I have thought about implementing a high level set of concepts outside of TP, but TP3 concepts are closely aligned to what I need. It also provides great tools (standard DSL) that I would like at a higher level anyway along with a nice architecture for implementing functionality in phases. It has moved a lot closer to implementing my data model out of the box versus in vertices and edges. 

I have read the discussions on the pros and cons of edge meta-properties. I have read multiple discussions on the pros and cons of hyper-edges and respect both sides of the argument. My point is rather than try and put forward a rigid model, put forward a default model (what you describe above) but also provide the ability for experienced developers to extend the data model that best suits their needs. 

Now, with respect to the practical side, here is why I'm looking for meta-properties on edges. I'm using the graph to store ontologies and then use them are part of the execution in realtime (it's a model of a model). We use properties on the relationships (edges) to help in the execution in realtime. We support security down to the property level and hence managing the authorization on relationship properties is a strong requirement.

For authorization on a property we model a set of actions and associated permissions needed to perform those actions on the property. We also store things like time created, modified, who changed it etc. In TP2 everything was stored in vertices. In TP3 vertex properties greatly simplify that. Considering a relationship as a higher level concept of an edge (in my world), as a workaround I'm implementing meta-properties on a relationship by implementing it as a vertex underneath. I'm now trying to make it appear like an Edge with meta-properties (vertex properties) in traversals. A purist would say that I should implement a new concept along with new DSL steps e.g. "edge-vertex" - EV) =>   g.EV().inV().outEV().properties()... etc. But considering there is so much code in TP3 already written that is close to supporting this I am trying to leverage the out-of-the-box functionality. 

The last practical use case relates to hyper-edges. I know that hyper-edges can be modeled with edges and vertices (which we did in TP2). But expanding on my authorization example above, we model a permission as a vertex. In TP2 we had an edge representing an action that referenced a permission. A traversal performing authorization would traverse the edges holding actions and pickup all the permissions. Our permissions were not static - they could be named whatever we wanted and we didn't have to worry about denormalization because they were stored in their own vertices versus storing their IDs as values on properties e.g. "read", "sensitive read" etc. Given this, when two domain objects are connected by a relationship, a hyper-edge representing the mapping for an action to a permission makes a lot of sense since it simplifies the modeling of the low-level concepts.

The trigger for this whole thread is because a number developers have started with vertex properties to implement authorization as advertised but have run into obstacles trying to implement the same for edges due to the lack meta-properties. You could argue these edges should be vertices - and you may be philosophically right - but if we did this then we're back in TP2 land.  

 Hopefully this was insightful.

Thanks,
Dave

**Comment by rjbriody** [comment]
@Matthias, very solid post. I think our philosophical viewpoints are very closely aligned.

>Vertices represent entities in a domain and edges represent a relationship between pairs of entities.

Agreed.

>Your entities should be vertices, your binary relationships edges and vertex-properties". 

Cool.

>first class: vertex. Second class: edge. Vertex properties are also second class citizens as they represent a relationship between a vertex and a value (again, a binary relationship). And there are third class citizens: key-value annotations on vertex-properties and edges.

I might even be able to get down w/ that.

Reading your post helped me realize something interesting: the challenging cases Bryn and I have run into with TP3 all involve concepts outside the core graph schema - meaning things like history, write auditing (what user wrote a value), and permissions. I find this interesting because **these concepts seem orthogonal to the philosophical discussion of core schema modeling**. 

I don’t think this is a big surprise. [Permissions and auditing are 2 of the driving examples for VertexProperties](http://www.tinkerpop.com/docs/3.0.0-SNAPSHOT/#vertex-properties). (I would throw history/versioning in that group as well.) These meta/secondary/whatever-you-want-to-call-it use cases seem equally applicable to Edge properties.

Getting pragmatic...

>So, a common approach would be to soft-delete the old edge and add a new edge with updated provenance information on it.

This works, but storing property history with multi/meta properties is just soooo much nicer. It's nicer to understand, implement, and query.

Regardless, having immutable edges still does not solve the requirements for write auditing or permissions on Edge properties.  For example: 

```Person -Knows-> Person``` 
```Knows``` has properties ```since``` and ```until``` 

A user might have permission to access the vertices and edges for "Bob knows Matthias" but may not have permission to see the ```since``` or ```until```" properties of that relationship. This is currently clunky to implement. Additionally, there is no natural place to capture the auditing data for what user entered the ```since``` or ```until``` property values, or the time of their creation/deletion.

**Comment by davidcrouch** [comment]
In thinking about Matthias' post and what we're trying to do with security,  I'm currently using a graph strategy to encapsulate a vertex with an "edge" - from a conceptual level I have an edge with meta-properties, which allows me to define authorization on an edge property. At the data level I have a vertex representing the edge, which is what Matthias was saying. So, the issue is how do I do something like this:-

g.E().properties("since").has("acl","public").grant("Dave")

We're looking in the wrong place. We should be focusing on improving the ability to enhance/override the standard DSL and/or easily adding our own custom DSL concepts since at certain point the data has to be written to disk according to a defined data model (again what Matthias is saying). 

As an example, let's assume a new concept "relationship", which is essentially an edge with meta-properties. A graph strategy could map it to a vertex in the graph DB. A traversal could be:-

g.R().properties("since").has("acl","public").grant("Dave")

Since the syntax using "R" (and related steps outR, inR etc.) is extremely close to "V", and a "relationship" is represented as a vertex in the graph DB, what is the best way to implement "R" by leveraging what has been done with "V"  (and related steps)?   That is, extending GraphTraversal  (I've been looking at Neo4jGraphTraversal)??

I think the ability to extend the standard DSL or define a custom one should be tightly integrated with graph strategy functionality since they seem to go hand-in-hand. To me this seems like the approach for modeling meta-properties on edges.   

Thoughts?

**Comment by mbroecheler** [comment]
Thank you for your insightful feedback, David.

>The trigger for this whole thread is because a number developers have started with vertex properties to implement authorization as advertised but have run into obstacles trying to implement the same for edges due to the lack meta-properties. You could argue these edges should be vertices - and you may be philosophically right - but if we did this then we're back in TP2 land.

I am thinking of the core TP3 data model here. I agree that your use case takes your application data model beyond the core TP3 data model, but I would argue that you can accommodate it therein. Your edges, with their properties on which you want to maintain provenance, should be vertices.
One of the challenges with the graph data model is that people adopt it because it is "whiteboard friendly", i.e. closely aligned with the conceptual model of the developer: "There is a parent and there is a child, we will make those vertices with a `parent` edge". That is very appealing since it makes data modeling easier and ideally gets you off the ground quicker (also taking into account here the query language support, ```v.inE('parent')``` is much easier than a join).
However, it is important to recognize when you leave that 'core' data model and go to that higher level as you point out. There is always a higher level of modeling. What if you wanted to ACL your permissions because you have a social network and your permission settings are guided by "higher level permission". Obviously, I could play that game recursively.
Most trivially, it already happens when you take the previous simple example and extend it to genealogy: Parenthood is now captured with place of birth, birth certificate, etc. That's when that simple edge turns into a vertex and one huge challenge that we face in the graph community is i) educating people on when that change occurs (so they don't get confused about the absence of edge meta-properties or start cramming things into edge properties a la "i'll just take that whole birth certificate and make it one serialized string property") and ii) provide excellent support for that transition and building higher level models on top.

> We're looking in the wrong place. We should be focusing on improving the ability to enhance/override the standard DSL and/or easily adding our own custom DSL concepts since at certain point the data has to be written to disk according to a defined data model.

That's exactly what I was trying to say when referencing "better support". The fact that you can define custom traversals in TP3 will hopefully be a game changer in making that possible and allow users to build higher level traversals with the same look-and-feel as the lower level ones.
This is the area where we probably need a lot more experimentation and feedback since that is very knew and much less thought through.

Have a look at the SocialTraversal that Marko implemented as a proof-of-concept and note that you can even overwrite common steps like ```outE```. If in your higher level data model all edges are effectively vertices, then what's the use of a low-level ```outE``` anyways? Your users (i.e. the users of the data model) should have to be concerned with such things - it'll just be you.
And then, we would be able to generalize such traversals and have a toolbox of "security traversals" that one can adopt for their particular use case.

Finally, I would like to come back to your statement:

> at certain point the data has to be written to disk according to a defined data model.

This is crucial from a core data model point of view. In databases in general, there is this concept of a record. Loosely speaking, this is the granularity at which data gets atomically modified. It is very important to fix what a record is because that's where a database can become smart about how data is laid out, modified, and queried. In other words, at some point the database needs to know what the atomic piece of data is that has no internal structure and specific bounds - that's the record.

In TP3, vertex-properties and edges are the record. A vertex is a collection of those records. This is great, because now graph databases know what to optimize for. If an edge could have meta-properties which get individually updated, then an edge is not a record, but a collection of records itself where a record is a vertex-property. So, from the perspective of the database, we just turned an edge into a vertex and it would have to maintain some sort of nested adjacency list which makes things a lot less efficient.

This is another way of arguing for the same thing: You need a core data model with simple concepts that can be efficiently represented and queried for in physical machines. Then you can build on top and introduce more complicated concepts.

**Comment by mbroecheler** [comment]
@rjbriody, I would argue that in those cases where you start adding permission to individual edge properties, those edges are effectively vertices.

Picking on the example you gave:
> A user might have permission to access the vertices and edges for "Bob knows Matthias" but may not have permission to see the since or until" properties of that relationship. This is currently clunky to implement. Additionally, there is no natural place to capture the auditing data for what user entered the since or until property values, or the time of their creation/deletion.

If a user can either see the `since` or `until` for all `knows` edges or none, you don't need vertex-properties. If this permission is given (or denied) on an individual edge basis, then I would argue that that edge is a vertex, because you are saying "for this particular friendship, user group X should not be able to see when they started knowing each other". I realize this is a hypothetical example, but I would continue to argue that an application data model which makes such fine grained statements about its logical world thinks of `friendship` as a first-class citizen, i.e. a vertex.
So, yes, you could argue that putting a permission property as a meta-property on `since` is nice and quick. I would argue that its dangerous because the requirement to do so is highly indicative of the fact that your friendship is really a first class citizen and sooner rather then later there will be a need to capture information like "how do we know those two are friends, when was that observed, what's our confidence in that observation, which analysts have reviewed this relationship and what was their conclusion?". In fact, that's exactly what I saw happen in a national security use case.

Bottom line: I would maintain that in 9x% of use cases, the need for meta-properties on edges indicates that those edges are really vertices. In that case, they should be treated as such and - if convenience in traversal is needed - TP3 new feature should be used to build a higher order traversal.

**Comment by BrynCooke** [comment]
Thank you to everyone for the insightful comments.
There have been so many that I am going to try and split some of it up so we don't lose anything.

### Vertices vs edges
@mbroecheler @rjbriody 
It looks like we all agree that edges != vertices.
The reason that people are treating edges as vertices is that the API is too similar.

SQL types in particular will look at an edge and think 'junction table'. Therefore perhaps we should consider making making edges immutable.
https://github.com/tinkerpop/tinkerpop3/issues/518

### Security/History strategies
It would be great if we could provide basic implementations of these, even if it is to just show how it can be achieved.

https://github.com/tinkerpop/tinkerpop3/issues/520
https://github.com/tinkerpop/tinkerpop3/issues/519

@davidcrouch @mbroecheler 
I had also been thinking that a combination of strategy and DSL would be a good basis for implementing these.

### Vertex Properties
For all the talk about the distinction between vertices and edges there is still one remaining issue was mentioned above, but I didn't explain it very well. Here is another go:

Say we have the following vertex where '~' indicated meta-property:

1. The user wants to store multiple groups that have read access:
```
{
   "name":"bryn" ~read:"group1", "group2"
}
```
This is not representable using vertex properties.

2. The user wants the history of meta-properties to be stored:
```
{
   "document":"REALLY_LARGE" ~read:"group1"
}
```
changes to 
```
{
   "document":"REALLY_LARGE" ~read:"group2"
}
```
The only way of representing this is 
```
{
   document:"REALLY_LARGE" ~read:"group1" ~timestamp:100
   document:"REALLY_LARGE" ~read:"group2" ~timestamp:200
}
```

It gets more complicated if the document property is a list in itself.

Having had a go at implementing a history and acl strategy the only way of making it work seemed to be to ban the client from using multi and meta-properties.

I suspect the answer is to represent some of this stuff as vertices rather than properties. In fact VertexProperties look like Vertices that have eager fetch on them.

Which makes me wonder: why not use the TP2 model but support multi and meta-properties?
The way that this could be achieved is that TP would define what collections and meta properties look like in terms of hints on the graph.

https://docs.google.com/presentation/d/15_cWP_ZAOWvmeRPEeeycjpkwdHOfNKFZ-pLMyEesy_k/edit?usp=sharing

Users would then be required to use a DSL to mutate and query these structures. This assumes that DSL are make available at an element level https://github.com/tinkerpop/tinkerpop3/issues/484

There are some advantages to this:
* Simpler API for TP, not just on traversals but also for implementing strategies.
* All graph implementations gain collection support. It is up to the individual graph vendors to optimise e.g. via eager fetch and or optimised data layout. Does this fix https://github.com/tinkerpop/tinkerpop3/issues/389?
* Unlimited nesting of collections.



Thoughts?

**Comment by mbroecheler** [comment]
@BrynCooke I did not fully understand the conclusions you draw from your section on vertex properties.

I understand that modeling history/acl on meta-properties causes the same issue it did for edges. If you want to maintain the history for changes to a permission setting on a vertex property then you also need this higher level structure to do so. As discussed with edges, you would then have to create a separate vertex. Where meta-properties help is when you want to have simple permission/history annotations on standard key-value properties. 

To reference back the old discussion on meta-properties: They are useful for first-order history/permission management but they are not the answer to all such use cases. In particular, when you move up the abstract level and what to support history on vertex-properties with meta-properties, then you also have to raise the level of abstraction as discussed for edges.
The point of meta-properties is just to have this atomic data element to build things on.

Now, you could build a simple permission graph that exposes a reduced TP2-style interface in that there are vertices with key-value pairs as properties and edges with immutable properties. You could then easily model permissions using meta-properties and properties on edges, i.e. you wouldn't need to wrap edges around vertices or any of the advanced strategies we discussed above and this simple data model with security would probably satisfy a great variety of use cases (its more expressive than accumulo). The point is, its up to the permission layer you wrap around the core data model to make such decisions and expose them to the user through a strategy/DSL traversal.
I don't think we need to make such decisions for the entire data model just to enable such higher level abstractions.

**Comment by davidcrouch** [comment]
@BrynCooke, I didn't really understand your section on Vertex Properties either. 

@mbroecheler - your last paragraph describes almost exactly what I coded and have running in TP2. It works, performance is OK, but there is a lot of code to maintain and it is now becoming cumbersome. But the design does give me flexibility - everything is a vertex and edges have almost nothing except a label. We also expose an edge as a property on the vertex.

In porting to TP3 I'm seeing a code reduction of about 60%-80% due to the new features, maybe more - THIS IS HUGE!!   So, I'm trying to push as much as possible into standard TP3 functionality as possible. 

"....expose them to the user through a strategy/DSL traversal"  

This is key to being able to support higher abstraction layers in an elegant way but I don't know if it is as achievable as to what is being implied (from my experiences so far). I think a big part of the problem is lack of experience with TP3 in creating this mapping/abstraction, hence the reasons for these long discussions. A real example proving such may just resolve the issues in this thread......   

@BrynCooke , do you want to compare notes since we're working on similar things?

**Comment by mbroecheler** [comment]
@davidcrouch:
> This is key to being able to support higher abstraction layers in an elegant way but I don't know if it is as achievable as to what is being implied (from my experiences so far). I think a big part of the problem is lack of experience with TP3 in creating this mapping/abstraction, hence the reasons for these long discussions. A real example proving such may just resolve the issues in this thread......

Agreed. This is more theory than practice on my part as well as I have not yet attempted to implement such higher order traversals in TP3. We are still very much in the weeds of trying to get things to work correctly. If you have the time/energy to trail-blaze here, that would be excellent.

**Comment by BrynCooke** [comment]
@mbroecheler 

>I understand that modeling history/acl on meta-properties causes the same issue it did for edges. If you want to maintain the history for changes to a permission setting on a vertex property then you also need this higher level structure to do so. 

Your comment fully captures my concerns about the the VertexProperties.

>Now, you could build a simple permission graph that exposes a reduced TP2-style interface in that there are vertices with key-value pairs as properties and edges with immutable properties. You could then easily model permissions using meta-properties and properties on edges, i.e. you wouldn't need to wrap edges around vertices or any of the advanced strategies we discussed above and this simple data model with security would probably satisfy a great variety of use cases 

Indeed this is what I ended up doing, but it comes with a cost. Not only must the client not use meta/multi-properties, but it also means the history strategy is unable to be layered with any other strategy that uses these features. The ACL strategy in particular had to use prefixes instead of meta-properties because of this.

Having said that it did fulfil the client requirements, and perhaps there is just no way of completely separating history and ACL strategies anyway.

@davidcrouch 
Yes, I can compare notes, I can be found on google hangouts.

**Comment by graben1437** [comment]
@BrynCooke , @davidcrouch , @mbroecheler
I hesitate to wade into such an weighty discussion at this point, but I wanted to explore the "history" part of this conversation to try to understand better.

This may come down to "what sort of history are we talking about here?"

With respect to the question of "history" - what is the role of the "transaction log" in your minds for recording "history" ?   From Wikipedia:  "In the field of databases in computer science, a transaction log (also transaction journal, database log, binary log or audit trail) is a history of actions executed..."

Titan 0.5.3 (maybe you are not using that), has a Transaction Log that  "can automatically log transactional changes for additional processing or as a record of change."   These logs can be processed, stored and queried.  I often put them in a separate store rather than putting them back into Titan.  These changes can be fed back into the graph as an "after effect" of a user transaction, if that is desired.

Are you using the transaction logs in your history tracking ?  Or is a history log good enough for many use cases ?

If a "User wants to be able to view the state of an element at any point in history. This applies to edges as well as vertices" , that history is fully available in the processed and stored transaction log including provenance of which user did it for individual items.

If the idea is to allow for traversals "as if it were 1999" inside the graph, aren't there existing temporal models to use with property graphs that avoid doing weird stuff to edges ?   

For history, I would have thought the discussion would have gone in the direction of asking for an "at()" step into TP3.  The "at()" step would specify a time filter for all vertices and edges, assuming some underlying properties are set correctly in the graph - such as "since" properties.

**Comment by mbroecheler** [comment]
@graben1437 "History" in the context of this conversation means capturing historical data as part of the data model and not the sequence of transactions that resulted in the current database state. As such, this is not about logging transactions but rather keeping those changes visible in the data model.
